### PR TITLE
engine_tirocks: add write batch and perf context

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?
Issue Number: Ref #13058 

What's Changed:

```commit-message
`PerfContext` is almost the same as rocks except reference type name
changes. `WriteBatch` is simplified by removing unnecessary fields
and optimized for merge batches.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/engine_tirocks/src/write_batch.rs [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+            for wb in &other.wbs[..=other.index] {
+                self_wb.append(wb).map_err(r2e)?;
+            }
+            return Ok(());
+        }
+        let self_wb = &mut self.wbs[self.index];
+        if self_wb.count() < WRITE_BATCH_LIMIT {
+            self_wb.append(&other.wbs[0]).map_err(r2e)?;
+        }
+        // From this point, either of following statements is true:
```
Comment (tabokie) [inline components/engine_tirocks/src/write_batch.rs:181]: What if `self_wb.count() >= WRITE_BATCH_LIMIT && other.index == 0`? other will be lost.

File: components/engine_tirocks/src/write_batch.rs [hunk 2]
```diff
@@ -0,0 +1,326 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use engine_traits::{Result, WriteBatchExt as _};
+use tirocks::{option::WriteOptions, WriteBatch};
+
+use crate::{r2e, RocksEngine};
+
+const WRITE_BATCH_MAX_BATCH: usize = 16;
+const WRITE_BATCH_LIMIT: usize = 16;
```
Comment (tabokie) [inline components/engine_tirocks/src/write_batch.rs:9]:
```suggestion
const WRITE_BATCH_MAX_BATCH_NUM: usize = 16;
const WRITE_BATCH_MAX_KEY_NUM: usize = 16;
```

File: components/engine_tirocks/src/write_batch.rs [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+            size += w.count();
+        }
+        size
+    }
+
+    fn is_empty(&self) -> bool {
+        self.wbs[0].as_bytes().is_empty()
+    }
+
+    fn should_write_to_engine(&self) -> bool {
```
Comment (tabokie) [inline components/engine_tirocks/src/write_batch.rs:115]: deserve inline.

File: components/engine_tirocks/src/perf_context.rs [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+        db_mutex_lock_nanos,
+    }
+
+    pub struct PerfContextTimeDuration : LocalHistogram {
+        "type" => PerfContextType
+    }
+}
+
+lazy_static! {
+    pub static ref APPLY_PERF_CONTEXT_TIME_HISTOGRAM: HistogramVec = register_histogram_vec!(
```
Comment (Connor1996) [inline components/engine_tirocks/src/perf_context.rs:56]: Why define these metrics in `tirocks`?

Comment (BusyJay) [reply components/engine_tirocks/src/perf_context.rs:56]: It's defined in `engine_rocks`. I guess it's a partial work to hide the details of perf context.

File: components/engine_tirocks/src/perf_context.rs [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+                for token in trackers {
+                    GLOBAL_TRACKERS.with_tracker(*token, |t| perf_context.report_to_tracker(t));
+                }
+                self.read += perf_context;
+                self.flush_read_metrics();
+            }
+        }
+    }
+
+    fn flush_read_metrics(&mut self) {
```
Comment (tabokie) [inline components/engine_tirocks/src/perf_context.rs:331]:
```suggestion
    fn maybe_flush_read_metrics(&mut self) {
```

**Review by tabokie — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]

**Pull request merged**